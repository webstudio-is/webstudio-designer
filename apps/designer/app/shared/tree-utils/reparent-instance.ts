import { type Instance } from "@webstudio-is/react-sdk";
import { findInstanceById } from "./find-instance";
import { findParentInstance } from "./find-parent-instance";
import { deleteInstanceMutable } from "./delete-instance";
import { insertInstanceMutable } from "./insert-instance";

export const reparentInstanceMutable = (
  rootInstance: Instance,
  instanceId: Instance["id"],
  newParentId: Instance["id"],
  newPosition: number
) => {
  const instance = findInstanceById(rootInstance, instanceId);
  if (instance === undefined) {
    return;
  }

  let newPositionAdjusted = newPosition;

  // If parent is the same, we need to adjust the position to account for the removal of the instance.
  const currentParent = findParentInstance(rootInstance, instance.id);
  if (currentParent !== undefined && currentParent.id === newParentId) {
    const currentPosition = currentParent.children.findIndex(
      (x) => typeof x !== "string" && x.id === instance.id
    );
    if (currentPosition < newPosition) {
      newPositionAdjusted--;
    }
  }

  deleteInstanceMutable(rootInstance, instance.id);
  insertInstanceMutable(rootInstance, instance, {
    parentId: newParentId,
    position: newPositionAdjusted,
  });
};
