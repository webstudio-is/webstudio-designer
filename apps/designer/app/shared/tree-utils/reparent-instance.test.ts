import { type Instance } from "@webstudio-is/react-sdk";
import { reparentInstanceMutable } from "./reparent-instance";

describe("Reparent instance", () => {
  test(`new parent`, () => {
    const rootInstance: Instance = {
      component: "Box",
      id: "1",
      cssRules: [],
      children: [
        {
          component: "Box",
          id: "2",
          cssRules: [],
          children: [],
        },
        {
          component: "Box",
          id: "3",
          cssRules: [],
          children: [
            {
              component: "Box",
              id: "4",
              cssRules: [],
              children: [],
            },
          ],
        },
      ],
    };

    reparentInstanceMutable(rootInstance, "4", "2", 0);
    expect(rootInstance).toMatchSnapshot();
  });

  test(`same parent different position`, () => {
    const rootInstance: Instance = {
      component: "Box",
      id: "1",
      cssRules: [],
      children: [
        {
          component: "Box",
          id: "2",
          cssRules: [],
          children: [],
        },
        {
          component: "Box",
          id: "3",
          cssRules: [],
          children: [],
        },
        {
          component: "Box",
          id: "4",
          cssRules: [],
          children: [],
        },
      ],
    };

    reparentInstanceMutable(rootInstance, "2", "1", 2);
    expect(rootInstance).toMatchSnapshot();
  });
});
