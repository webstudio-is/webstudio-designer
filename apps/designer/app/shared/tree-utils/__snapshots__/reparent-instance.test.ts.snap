// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reparent instance new parent 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "component": "Box",
          "cssRules": Array [],
          "id": "4",
        },
      ],
      "component": "Box",
      "cssRules": Array [],
      "id": "2",
    },
    Object {
      "children": Array [],
      "component": "Box",
      "cssRules": Array [],
      "id": "3",
    },
  ],
  "component": "Box",
  "cssRules": Array [],
  "id": "1",
}
`;

exports[`Reparent instance same parent different position 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "component": "Box",
      "cssRules": Array [],
      "id": "3",
    },
    Object {
      "children": Array [],
      "component": "Box",
      "cssRules": Array [],
      "id": "2",
    },
    Object {
      "children": Array [],
      "component": "Box",
      "cssRules": Array [],
      "id": "4",
    },
  ],
  "component": "Box",
  "cssRules": Array [],
  "id": "1",
}
`;
